File: virtual-machine/src/source/scanner.cpp
#include "../include/scanner.h"

#include <cstring>

Token Scanner::scanToken()
{
    skipWhitespacesAndComments();
    start = current;
    if (isAtEnd())
    {
        return makeToken(TokenType::FILE_EOF);
    }

    auto const c = advance();
    if (isLetterOrUnderscore(c))
    {
        return identifier();
    }

    if (isdigit(c))
    {
        return number();
    }

    switch (c)
    {
        // One char tokens
        case '(': return makeToken(TokenType::LEFT_PAREN);
        case ')': return makeToken(TokenType::RIGHT_PAREN);
        case '{': return makeToken(TokenType::LEFT_BRACE);
        case '}': return makeToken(TokenType::RIGHT_BRACE);
        case ':': return makeToken(TokenType::COLON);
        case ',': return makeToken(TokenType::COMMA);
        case '.': return makeToken(TokenType::DOT);
        case '-': return makeToken(TokenType::MINUS);
        case '+': return makeToken(TokenType::PLUS);
        case ';': return makeToken(TokenType::SEMICOLON);
        case '*': return makeToken(TokenType::STAR);
        case '%': return makeToken(TokenType::MODULO);
        case '^': return makeToken(TokenType::EXPONENT);
        case '/': return makeToken(TokenType::SLASH);

        // Either one or two char tokens
        case '!': return makeToken(match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
        case '=': return makeToken(match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
        case '>':
        {
            if (match('>'))
            {
                return makeToken(TokenType::RSHIFT);
            }

            return match('=') ? makeToken(TokenType::GREATER_EQUAL) : makeToken(TokenType::GREATER);
        }
        case '<':
        {
            if (match('<'))
            {
                return makeToken(TokenType::LSHIFT);
            }

            return match('=') ? makeToken(TokenType::LESS_EQUAL) : makeToken(TokenType::LESS);
        }

        // Literals
        case '"': return string();

        default: break;
    }

    return errorToken("Unexpected character");
}

bool Scanner::isAtEnd() const
{
    return current >= source.length();
}

bool Scanner::isLetterOrUnderscore(const char c)
{
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}

Token Scanner::makeToken(const TokenType type) const
{
    const auto lexeme = std::string_view{source.begin() + start, source.begin() + current};
    return Token{type, lexeme, line};
}

Token Scanner::errorToken(const std::string_view &message) const
{
    return Token{TokenType::ERROR, message, line};
}

char Scanner::advance()
{
    current++;
    return source[current - 1];
}

bool Scanner::match(const char expected)
{
    if (isAtEnd())
    {
        return false;
    }

    if (source[current] != expected)
    {
        return false;
    }

    current++;
    return true;
}

char Scanner::peek() const
{
    return source[current];
}

char Scanner::peekNext() const
{
    if (isAtEnd())
    {
        return '\0';
    }

    return source[current + 1];
}

Token Scanner::string()
{
    while (peek() != '"' && !isAtEnd())
    {
        if (peek() == '\n')
        {
            line++;
        }

        advance();
    }

    if (isAtEnd())
    {
        return errorToken("Unterminated string.");
    }

    advance(); // Consume closing quote
    return makeToken(TokenType::STRING);
}

Token Scanner::number()
{
    while (isdigit(peek()))
    {
        advance();
    }

    if (peek() == '.' && isdigit(peekNext()))
    {
        advance();
        while (isdigit(peek()))
        {
            advance();
        }
    }

    return makeToken(TokenType::NUMBER);
}

Token Scanner::identifier()
{
    while (isLetterOrUnderscore(peek()) || isdigit(peek()))
    {
        advance();
    }

    return makeToken(identifierType());
}

TokenType Scanner::identifierType() const
{
    const std::string_view lexeme(&source[start], current - start);

    if (lexeme == "and") return TokenType::AND;
    if (lexeme == "break") return TokenType::BREAK;
    if (lexeme == "class") return TokenType::CLASS;
    if (lexeme == "continue") return TokenType::CONTINUE;
    if (lexeme == "do") return TokenType::DO;
    if (lexeme == "else") return TokenType::ELSE;
    if (lexeme == "false") return TokenType::FALSE;
    if (lexeme == "for") return TokenType::FOR;
    if (lexeme == "fun") return TokenType::FUN;
    if (lexeme == "if") return TokenType::IF;
    if (lexeme == "import") return TokenType::IMPORT;
    if (lexeme == "let") return TokenType::LET;
    if (lexeme == "nand") return TokenType::NAND;
    if (lexeme == "nor") return TokenType::NOR;
    if (lexeme == "null") return TokenType::NIL;
    if (lexeme == "or") return TokenType::OR;
    if (lexeme == "print") return TokenType::PRINT;
    if (lexeme == "return") return TokenType::RETURN;
    if (lexeme == "static") return TokenType::STATIC;
    if (lexeme == "super") return TokenType::SUPER;
    if (lexeme == "this") return TokenType::THIS;
    if (lexeme == "true") return TokenType::TRUE;
    if (lexeme == "while") return TokenType::WHILE;
    if (lexeme == "xor") return TokenType::XOR;

    return TokenType::IDENTIFIER;
}

void Scanner::skipWhitespacesAndComments()
{
    for (;;)
    {
        switch (peek())
        {
            case ' ':
            case '\r':
            case '\t':
                advance();
                break;
            case '\n':
                line++;
                advance();
                break;
            case '/':
                if (peekNext() == '/')
                {
                    while (peek() != '\n' && !isAtEnd())
                    {
                        advance();
                    }
                }
                else if (peekNext() == '*')
                {
                    advance(); // Consume the opening /
                    advance(); // Consume the opening *

                    while (!isAtEnd() && (peek() != '*' || peekNext() != '/'))
                    {
                        if (peek() == '\n')
                        {
                            line++;
                        }

                        advance();
                    }

                    advance(); // Consume the closing *
                    advance(); // Consume the closing /
                }
                else
                {
                    return;
                }

                break;
            default:
                return;
        }
    }
}

File: lang/src/main/kotlin/core/scanner/Scanner.kt
package core.scanner

import core.`object`.Keywords
import core.enum.TokenType
import core.error.reporter.ErrorReporter


class Scanner(
    val source: String,
    val errorReporter: ErrorReporter,
    private val tokens: MutableList<Token> = mutableListOf(),
    private var start: Int = 0,
    private var current: Int = 0,
    private var line: Int = 0,
) {
    fun scanTokens(): List<Token> {
        while (!isAtEnd()) {
            start = current
            scanToken()
        }

        tokens.add(Token(TokenType.EOF, "", null, line))
        return tokens
    }

    private fun isAtEnd(): Boolean {
        return current >= source.length
    }

    private fun scanToken() {
        when (val char = advance()) {
            '(' -> addToken(TokenType.LEFT_PAREN)
            ')' -> addToken(TokenType.RIGHT_PAREN)
            '{' -> addToken(TokenType.LEFT_BRACE)
            '}' -> addToken(TokenType.RIGHT_BRACE)
            '[' -> addToken(TokenType.LEFT_SQUARE_BRACKET)
            ']' -> addToken(TokenType.RIGHT_SQUARE_BRACKET)
            ',' -> addToken(TokenType.COMMA)
            '.' -> addToken(TokenType.DOT)
            '+' -> addToken(TokenType.PLUS)
            '-' -> addToken(TokenType.MINUS)
            '%' -> addToken(TokenType.MODULO)
            '^' -> addToken(TokenType.EXPONENT)
            ';' -> addToken(TokenType.SEMICOLON)
            ':' -> addToken(TokenType.COLON)
            '!' -> addToken(if (match('=')) TokenType.BANG_EQUAL else TokenType.BANG)
            '=' -> addToken(if (match('=')) TokenType.EQUAL_EQUAL else TokenType.EQUAL)
            '>' -> addToken(
                when {
                    match('=') -> TokenType.GREATER_EQUAL
                    match('>') -> TokenType.RSHIFT
                    else -> TokenType.GREATER
                }
            )

            '<' -> addToken(
                when {
                    match('=') -> TokenType.LESS_EQUAL
                    match('<') -> TokenType.LSHIFT
                    else -> TokenType.LESS
                }
            )

            '*' -> {
                if (isInMultilineComment && match('/')) {
                    isInMultilineComment = false
                } else {
                    addToken(TokenType.STAR)
                }
            }

            '/' -> {
                if (match('*')) {
                    // Multiline comments : /**/
                    isInMultilineComment = true
                    while ((peek() != '*' || peekNext() != '/') && !isAtEnd()) {
                        advance()
                    }

                    if (!isAtEnd()) {
                        advance()
                        advance()
                        isInMultilineComment = false
                    }
                } else if (match('/')) {
                    // Single line comments : //
                    while (peek() != '\n' && !isAtEnd()) advance()
                } else {
                    addToken(TokenType.SLASH)
                }
            }

            '"' -> string()
            ' ', '\r', '\t', '\n' -> {}
            'o' -> {
                if (match('r')) {
                    addToken(TokenType.OR)
                }
            }

            else -> {
                if (isDigit(char)) {
                    number()
                } else if (isAlpha(char)) {
                    identifier()
                } else if (!isInMultilineComment) {
                    errorReporter.report(line, "Unexpected character : $char")
                }
            }
        }
    }

    private fun match(expected: Char): Boolean {
        if (isAtEnd()) {
            return false
        }

        if (source[current] != expected) {
            return false
        }

        current++
        return true
    }

    private fun isDigit(char: Char): Boolean {
        return char in '0'..'9'
    }

    private fun isAlpha(char: Char): Boolean {
        return (char in 'a'..'z') || (char in 'A'..'Z') || char == '_'
    }

    private fun isAlphaNumeric(char: Char): Boolean {
        return isAlpha(char) || isDigit(char)
    }

    private fun advance(): Char {
        current++
        if (source[current - 1] == '\n') {
            line++
        }

        return source[current - 1]
    }

    private fun peek(): Char {
        return if (isAtEnd()) 0.toChar() else source[current]
    }

    private fun peekNext(): Char {
        if (current + 1 >= source.length) return 0.toChar()
        return source[current + 1]
    }

    private fun string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n') line++
            advance()
        }

        if (isAtEnd()) {
            errorReporter.report(line, "Unterminated string.")
        }

        advance()

        val strValue = source.substring(start + 1, current - 1)
        addToken(TokenType.STRING, strValue)
    }

    private fun number() {
        while (isDigit(peek())) advance()

        if (peek() == '.' && isDigit(peekNext())) {
            advance()
            while (isDigit(peek())) advance()
        }

        addToken(TokenType.NUMBER, source.substring(start, current).toDouble())
    }

    private fun identifier() {
        while (isAlphaNumeric(peek())) advance()
        val text = source.substring(start, current)
        val tokenType = Keywords.all[text] ?: TokenType.IDENTIFIER
        addToken(tokenType)
    }

    private fun addToken(type: TokenType, literal: Any? = null) {
        if (!isInMultilineComment) {
            val text = source.substring(start, current)
            tokens.add(Token(type, text, literal, line))
        }

        // Just ignore everything
    }

    companion object {
        private var isInMultilineComment: Boolean = false
    }
}